pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -274,61 +363,154 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
 	p.removeRestoredRulesLocked(uint64(endpointID))
 }
 
+// lookupOrCompilePatternMatcher will check if the pattern is already compiled and present in another policy, and
+// will reuse it in order to reduce memory consumption. The usage is reference counted, so all calls where
+// lookupOrCompilePatternMatcher returns no error, a subsequent call to release it via releasePatternMatcher has to
+// be tone when its no longer being used by the policy.
+func (c *matcherCache) lookupOrCompilePatternMatcher(pattern string) (*regexp.Regexp, error) {
+	if pattern == "" {
+		return nil, nil
+	}
+	if entry, ok := c.patternMatcherCache[pattern]; ok {
+		entry.referenceCount += 1
+		return entry.matcher, nil
+	}
+	regex, err := re.CompileRegex(pattern)
+	if err != nil {
+		return nil, err
+	}
+	c.patternMatcherCache[pattern] = &patternMatcherEntry{matcher: regex, referenceCount: 1}
+	return regex, nil
+}
+
+// lookupOrInsertPatternMatcher is equivalent to lookupOrCompilePatternMatcher, but where a compiled regex is provided
+// instead of the pattern. In case a compiled regex with the same pattern as the provided regex is already present in
+// the cache, the already present regex will be returned. By doing that, the duplicate can be garbage collected in case
+// there are no other references to it. Trying to insert a nil value is a noop and will return nil
+func (c *matcherCache) lookupOrInsertPatternMatcher(regex *regexp.Regexp) *regexp.Regexp {
+	if regex == nil {
+		return nil
+	}
+	pattern := regex.String()
+	if entry, ok := c.patternMatcherCache[pattern]; ok {
+		entry.referenceCount += 1
+		return entry.matcher
+	}
+	c.patternMatcherCache[pattern] = &patternMatcherEntry{matcher: regex, referenceCount: 1}
+	return regex
+}
+
+// releasePatternMatcher releases the pattern matcher. In case there are no longer any references to it,
+// it will be freed. Running release on a nil value is a noop.
+func (c *matcherCache) releasePatternMatcher(patternMatcher *regexp.Regexp) {
+	if patternMatcher == nil {
+		return
+	}
+	if indexEntry, ok := c.patternMatcherCache[patternMatcher.String()]; ok {
+		switch indexEntry.referenceCount {
+		case 1:
+			delete(c.patternMatcherCache, patternMatcher.String())
```

Comment:
Let's call `String()` only once so we don't waste cycles doing duplicate work

Reply:
Yeah, sounds good!